File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.Map;
import java.util.PriorityQueue;

class Solution {
    public String solution(String m, String[] musicinfos) {
        Map<String, String> map = Map.of("C#","c","D#","d","F#","f","G#","g","A#","a","B#","b");
        PriorityQueue<Song> pq = new PriorityQueue<>((s1, s2) -> Integer.compare(s2.playTime, s1.playTime));
        String temp = m;

        for (String key : map.keySet()) {
            temp = temp.replace(key, map.get(key));
        }

        for (String info : musicinfos) {
            pq.add(new Song(info, map));
        }

        while (!pq.isEmpty()) {
            Song song = pq.poll();

            if(song.notes.contains(temp)){
                return song.songName;
            }
            
        }
        return "(None)";
    }
}

class Song{
    int startTime;
    int endTime;
    int playTime;
    String songName;
    String notes;

    private Song() {;}

    public Song(String info, Map<String, String> map) {
        String[] infos = info.split(",");
        this.startTime = parseTime(infos[0]);
        this.endTime = parseTime(infos[1]);
        this.playTime = this.endTime - this.startTime;
        this.songName = infos[2];


        for (String key : map.keySet()) {
            infos[3] = infos[3].replace(key, map.get(key));
        }

        StringBuilder sb = new StringBuilder();
        for(int i = 0; i < this.playTime; i++){
            sb.append(infos[3].charAt(i % infos[3].length()));
        }
        this.notes = sb.toString();
    }
    
    private int parseTime(String time){
        String[] times = time.split(":");
        return Integer.parseInt(times[0]) * 60 + Integer.parseInt(times[1]);
    }
}

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
def solution(m, musicinfos):
    answer = ''
    note_index = {
        'C#':'b','D#':'d','F#':'g', 'G#':'i','A#':'k','B#':'m','E#':'n',
        'C':'a','D':'c','E':'e','F':'f','G':'h','A':'j','B':'l'
    }
    
    def time_2_int(time:str)->int:
        return 60*int(time[:2]) + int(time[3:])
    
    def parse_music(music:str)->str:
        result = ''
        i = 0
        while i < len(music):
            if i + 1 < len(music) and music[i+1] == '#':
                result += note_index[music[i:i+2]]
                i += 2
                continue
            result += note_index[music[i]]
            i += 1
        return result
    
    m = parse_music(m)
    max_play_time = 0
    for info in musicinfos:
        start_time, end_time, title, music = info.split(',')
        play_time = time_2_int(end_time) - time_2_int(start_time)
        music = parse_music(music)
        if play_time > len(music):
            music *= play_time // len(music) + 1
        if m in music[:play_time] and max_play_time < play_time:
            answer = title
            max_play_time = play_time
    
    if max_play_time == 0:
        return "(None)"
    
    return answer